### ISSUE TITLE
Linear solver benchmarking on EqQPs

### ISSUE THREAD

**Issue by GeoffNN** [google/jaxopt#168]
I benchmarked GMRES (currently default in jax.eq_qp) vs other `scipy` solvers: [minres](https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.linalg.minres.html#scipy.sparse.linalg.minres) and [LGMRES](https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.linalg.lgmres.html#scipy.sparse.linalg.lgmres). I sampled random equality constrained QP KKT matrices, and found pretty stark differences between the solvers. Primal dim: 1500, dual dim: 1000.

TLDR: LGMRES is much faster, and GMRES is the slowest of the three. NB: Minres requires the matrix in the linear system to be symmetric (although indefinite), while gmres and lgmres don't.

I recall a discussion about linear solvers living in the `jaxopt` package vs `jax.scipy.sparse.linalg`. 
I don't have time right now to implement LGMRES, (and would be grateful if anyone has the bandwidth and the interest) but I will try to get around to it in the next few weeks.  Should the solver's code live in jaxopt?

Colab for repro: https://colab.research.google.com/drive/1Ge1-gmuknDQq0rHpnSrHpvwLbU23i6YG?usp=sharing

![image](https://user-images.githubusercontent.com/11814989/152444223-57ba6bea-0c2c-4eb4-aa62-afa9ab70e575.png)

**Comment by choltz95** [comment]
I don't really have anything useful to add, but I found that cg and bicg were superior wrt to stability & much better wrt runtime on large sparse and large sparse "slightly singular" systems- quadratics involving a graph Laplacian. 

 I had little success preconditioning any of them

**Comment by GeoffNN** [comment]
Yeah, the reason I'm not using CG is because the EqQP is indefinite -- so I'd have to use CG on the normal system `A.T A = A.T b`, which can have a really big condition number.

**Comment by GeoffNN** [comment]
The restart parameter for GMRES is critical. Here's a new plot with `restart=625` for this example.

LGMRES is an automated way of finding the `restart` parameter, amongst other things).

 Perf for GMRES is now on par with LGMRES. 
 

![image](https://user-images.githubusercontent.com/11814989/152598567-e0ebc18c-8fc2-4c99-af81-aa3c0d4e371d.png)

**Comment by mblondel** [comment]
Interesting results! I would be also interested in a benchmark of JAX gmres vs. SciPy gmres.

**Comment by GeoffNN** [comment]
@mblondel sorry I didn't get around to that benchmark. I recently found [this JAX issue](https://github.com/google/jax/issues/9259) which says that GMRES is much slower on GPU than on CPU for Jax. 
I'm now wondering how to wrap the GMRES to move things to CPU, including for the backward pass, if it's used as a submodule in a longer differentiable pipeline.
